File: C++/Medium/Accepted/evaluate-the-bracket-pairs-of-a-string_2.cpp
class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        
        unordered_map<string, string> m;
        for (auto& v : knowledge) {
            m[v[0]] = v[1];
        }
        string res = "";
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                int j = i + 1;
                while (j < s.size() && s[j] != ')') j++;
                if (j < s.size()) {
                    string k = s.substr(i + 1, j - i - 1);
                    if (m.find(k) != m.end()) {
                        res += m[k];
                    } else {
                        res += "?";
                    }
                    i = j;
                } else {
                    res += "?";
                }
            } else {
                res += s[i];
            }
        }
        return res;
    }
};

File: Rust/Medium/WrongAnswer/evaluate-the-bracket-pairs-of-a-string_2.rs
impl Solution {
    pub fn evaluate(s: String, knowledge: Vec<Vec<String>>) -> String {
        
        let mut map = std::collections::HashMap::new();
        for i in knowledge {
            map.insert(i[0].clone(), i[1].clone());
        }
        let mut res = String::new();
        let mut i = 0;
        while i < s.len() {
            if s[i..].starts_with('(') {
                let mut j = i + 1;
                while s[j..].starts_with('(') {
                    j += 1;
                }
                let mut k = j + 1;
                while s[k..].starts_with('(') {
                    k += 1;
                }
                res.push_str(map.get(&s[j..k]).unwrap_or(&"?".to_string()));
                i = k + 1;
            } else {
                res.push(s[i..].chars().next().unwrap());
                i += 1;
            }
        }
        res
    }
}